src/SquidConfig.h

```diff
@@ -242,7 +243,7 @@ class SquidConfig
     size_t tcpRcvBufsz;
     size_t udpMaxHitObjsz;
     wordlist *mcast_group_list;
-    CachePeer *peers;
+    CachePeers *peers;
     int npeers;
```

Comment:
You appear to have a `CurrentCachePeers()` static global that provides the appropriate API for accessing the configuration container. Please replace these `SquidConfig::peers` and `SquidConfig::npeers` members with a static container inside the `CurrentCachePeers()` function.

Reply:
> You appear to have a CurrentCachePeers() static global that provides the appropriate API for accessing the configuration container

CurrentCachePeers() is a global convenience function that provides safe access to code that wants to work with a _constant_ container of still-available configured cache peers without worrying about their existence. It is essential for using C++ range-based for loops (which usually improve code quality but are not appropriate for all use cases).

> Please replace these SquidConfig::peers and SquidConfig::npeers members with a static container inside the CurrentCachePeers() function.

It is not possible to use CurrentCachePeers() to store SquidConfig::peers because CurrentCachePeers() provides read-only access. We can, of course, move SquidConfig::peers to a static variable inside CachePeers.cc and provide new functions to write that variable, but I believe all of that work should be declared outside _this_ minimal PR scope.

Removing SquidConfig::peers would create significant out-of-scope PR noise and complications. Even cf.data.pre would need to be modified, and if we start modifying that, I can imagine many serious improvements that suddenly become in-scope! This _minimal_ PR improves SquidConfig::peers _type_ (upgrading it from a raw invasive list to a proper container).  This PR should not have to _also_ change where SquidConfig::peers are stored. To keep this PR focused/minimal, any old code that can successfully use Config.peers should continue to do so in _this_ PR.

Reply:
Hmm, that is wrong. The type change brings an API change which has already required a change to all code lines using the config object.
The only additional change I see from this request is logic using+maintaining the now unnecessary `SquidConfig::npeers`  value.

Reply:
> The only additional change I see from this request is logic using+maintaining the now unnecessary `SquidConfig::npeers` value.

Satisfying this change request would also require creating new code/logic to store/maintain a place where configured cache peers are stored and manipulated. That change, in turn, brings significant cache_peer parser changes into this PR scope (at least!). All that work is currently outside this PR scope. The benefits of moving that storage away from Config are also debatable, but I am trying to avoid that out-of-scope debate.

If we really want minimal PRs, we should stop insisting on changes that _can_ be considered outside a reasonable PR scope, even if there is strong consensus that those changes represent a step forward. We will improve the affected cache peer storage and parsing code in the foreseeable future -- this PR is one of dozens PRs in the project that, collectively, solves many long-term configuration problems. Please let us move forward using these small but well-scoped steps.

---

> Hmm, that is wrong. 

FWIW, I do not know what "that" is in this context, so I am addressing the other/specific comments only.

> The type change brings an API change which has already required a change to all code lines using the config object.

Even the above "all code lines" assertion is incorrect: Not all code lines using Config.peers required a change in this PR. For example, some of the lines testing whether cache peers were configured at all remained as is. There are other use cases where the type change did not matter.

Reply:
IMO this request is part of the policy change we agreed on in todays developer meeting. About preferring not to use globals. That was the intention behind this request.
Since we now have a formal policy about it and `SquidConfig` is a pre-existing global I will accept shifting this request to a followup PR.

Reply:
Thank you for clearing this PR!

I would like to clarify one aspect in hope to reduce future conflicts:

> IMO this request is part of the policy change we agreed on in todays developer meeting. About preferring not to use globals

The "no new globals" policy does not apply to Config.peers because a non-static data member is not a "global". The policy does not even apply to Config itself because that global is grandfathered (i.e. not "new"). Thus, this and similar change requests should not be viewed as a "part of [that] policy change".

Config.peers field may stay the same or may change. If it changes, it will not be because of the "no new globals" policy. Same for the other Config data members.

Config global may change because of the intent behind the "no new globals" policy. Whether it will be eventually replaced with CurrentConfig() or just adjusted to rely on static zero-initialization of its POD data members (individual data member declaration bugs notwithstanding) remains to be seen.